Use FeatureSwitchDefinition in a few places

sbomer opened the pull request:
This replaces much of our ILLink.Substitutions.xml with nearly equivalent FeatureSwitchDefinitionAttribute. The minor difference is that both true and false are substituted (when the feature switch is set accordingly), whereas some of our substitution xmls only defined a substitution for the false case.

sbomer committed:
Use FeatureSwitchDefinition in a few places

jkoritzinsky reviewed (commented):
Interop-related switches LGTM

sbomer committed:
Fix project files

sbomer committed:
Missed one

sbomer committed:
Fix native AOT feature switches

- IsReflectionExecutionAvaliable shouldn't have used FeatureSwitchDefinitionAttribute
  because it's not a property
- Added missing FeatureSwitchDefinitionAttribute to MetadataUpdater.IsSupported

sbomer committed:
Cleanup

sbomer committed:
Fix System.Text.Json build

sbomer committed:
Fix allconfigurations libraries build

sbomer committed:
Merge remote-tracking branch 'origin/main' into useFeatureSwitchAttributes

jtschuster commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeType.cs:
```diff
@@ -46,6 +46,7 @@ internal void Free()
             RuntimeImports.RhHandleFree(_runtimeTypeInfoHandle);
         }
 
+        [FeatureSwitchDefinition("System.Reflection")]
         private static bool IsReflectionDisabled => false;
```
Doesn't this property reflect the negated value of that feature switch? Does this account for that somehow?

```xml
  <type fullname="System.RuntimeType" feature="System.Reflection" featurevalue="false">
    <method signature="System.Boolean get_IsReflectionDisabled()" body="stub" value="true" />
  </type>
```

MichalStrehovsky replied:
> No, great catch, that one can't be handled by FeatureSwitchDefinition. I'm surprised this didn't cause test failures in ci. @MichalStrehovsky

We don't test things with reflection disabled since that mode is unsupported. So if this enabled reflection unconditionally, I'm not surprised it wasn't caught. I'd be surprised if it was the other way and nothing failed.

sbomer committed:
Revert System.Reflection FeatureSwitchDefinition

jtschuster commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs:
```diff
@@ -37,6 +37,7 @@ public sealed class ServiceProvider : IServiceProvider, IKeyedServiceProvider, I
         internal static bool VerifyOpenGenericServiceTrimmability { get; } =
```
Is this missing from src/libraries/Microsoft.Extensions.DependencyInjection/src/ILLink/ILLink.Substitutions.xml?

```suggestion
        [FeatureSwitchDefinition("Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability")]
        internal static bool VerifyOpenGenericServiceTrimmability { get; } =
```

jtschuster commented on src/libraries/System.Text.Json/src/System.Text.Json.csproj:
```diff
@@ -369,6 +365,8 @@ The System.Text.Json library is built-in as part of the shared framework in .NET
     <Compile Include="$(CommonPath)System\Collections\Generic\EnumerableHelpers.cs" Link="Common\System\Collections\Generic\EnumerableHelpers.cs" />
     <Compile Include="$(LibrariesProjectRoot)System.Collections\src\System\Collections\Generic\OrderedDictionary.cs" Link="Common\System\Collections\Generic\OrderedDictionary.cs" />
     <Compile Include="$(LibrariesProjectRoot)System.Collections\src\System\Collections\ThrowHelper.cs" Link="Common\System\Collections\ThrowHelper.cs" />
+    
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\FeatureSwitchDefinitionAttribute.cs" />
```
nit: remove extra line
```suggestion
    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\FeatureSwitchDefinitionAttribute.cs" />
```

jtschuster reviewed (approved):
LGTM modulo comments, thank you!

sbomer committed:
Update src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs

Co-authored-by: Jackson Schuster <36744439+jtschuster@users.noreply.github.com>

eerhardt commented on src/libraries/System.Text.Json/ref/System.Text.Json.csproj:
```diff
@@ -24,6 +24,10 @@
     <Compile Include="$(CoreLibSharedDir)System\Runtime\Versioning\RequiresPreviewFeaturesAttribute.cs" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '9.0'))">
```
```suggestion
  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0')">
```

?

eerhardt reviewed (commented):
Can we do the System.SR one as well (obviously in a separate PR)?

https://github.com/dotnet/runtime/blob/269ea267270838478e15e29f51c250f9bea81b67/src/libraries/Common/src/System/SR.cs#L10-L13
